### PR TITLE
mailbox: Minimize messages latency

### LINKED ISSUE #102: Minimize storage mailbox latency
Sending messages using the storage mailbox has very high latency
of 2-4 seconds. This can cause VMs to pause when they wait for
extension request completion.

## Why to we have 2-4 seconds latency?

On the SPM host, the SPM mailbox monitor reads the SPM inbox every
2 seconds. Every host has a 4k mailbox in the SPM inbox. We support
up to 250 hosts in a DC, so the SPM has to read the mailbox for all
host - 250 * 4k -> 1000 KiB. So the SPM is reading 500 KiB/s from
the ibox.

On the other hosts, every host has a HSM mailbox monitor. This monitor
read the host inbox every 2 seconds, when the host is waiting from 
reply from the SPM. Since mailbox size is 4k, the host read 2k/s during
extension request.

## Extension request lifetime

1. Host send extension request, writing 4k to the host outbox
2. The SPM discover the host request 0-2 seconds later
3. The SPM extend the disk and write a reply to the host inbox
4. The host discover the SPM reply 0-2 seconds later
5. The host refresh the extended disk

From this we can see that in the worst case, both the SPM and
the host will wait 2 seconds, and the entire roundtrip will take
at least 4 seconds, regardless of the time to extend the volume
(typically 0.2 seconds).

### Reduce the check interval on the SPM and other hosts

Change the check internal to 0.5-1.0 seconds. This will increase the I/O
specially on the SPM host, which reads almost 1 MiB on every read.

Keeping the current code, I think we can decrease the check interval to 
1 second on the SPM, and 0.1 seconds on other hosts since they read
only 4k.

### Introduce an event block

Instead of reading the entire inbox (100k), the SPM can read one
block (4k) used only for signaling changes. We can use the mailbox at
offset 0 for this, since host id 0 is reserved and not used by any
host (enforced by sanlock).

The SPM will monitor the event block, and when it detect that the event
block has changed, it will read the entire inbox to find new message
from all hosts in the DC.

Since the SPM read one block instead of entire inbox, we can use much
lower check internal, for example 0.1 seconds. This can decrease the 
latency to 0.1-0.2 seconds.

### Detailed design

After a host write a message to the the SPM inbox, it will write an
event to the event block in the SPM inbox. The first byte of the
event must be 0, to be compatible with older hosts - they will skip
this message in the mailbox.
https://github.com/oVirt/vdsm/blob/3b87510e20a45a5d5d17cd57f5fe5d475ceb2179/lib/vdsm/storage/mailbox.py#L627

Since many hosts can write to the event block at the same time, the SPM
will see only the last write. This is correct, since the purpose of
the event is to signal the SPM host to read the entire mailbox.

To make this work, event must have unique content. An easy way to 
to implement this is to include a random UUID in the event.

The SPM will monitor the event block with a smaller interval. For example
it can read the event block 10 times per second. This will issue 10 
4k reads per second, decreasing the latency to 0.1 seconds, and the amount
of data read from 500k/s to 40k/s.

On the hosts, the HSM mail monitor will also read the host inbox using
smaller interval. If we use 0.1 seconds internal, the latency will decrease
to 0.1 seconds, but the amount of data read will increase from 2k/s to 40k/s.

#### Latency vs I/O

| interval (s) | read (k/s) |
|--------------|------------|
| 0.1          |         40 |
| 0.2          |         20 |
| 0.5          |          8 |
| 1.0          |          4 |

#### Event block format

|    range    |   content                              |
|-------------|----------------------------------------|
| 0-4         | b"\0evt" (first byte must be zero)     |
| 4-20        | UUID bytes in big endian order         |
| 20-4096     | padding                                |

Example - formatting an event:

```
>>> import uuid
>>> buf = bytearray(4096)
>>> buf[0:4] = b"\0evt"
>>> buf[4:20] = uuid.uuid4().bytes
>>> buf[:20]
bytearray(b'\x00evt\x12\x87\x98\t\xd6\xbaF\xc7\xbdie\x0b\xa8\xe0\xd0\x9a')
```

Example parsing an event:

```
>>> uuid.UUID(bytes=bytes(buf[4:20]))
UUID('12879809-d6ba-46c7-bd69-650ba8e0d09a')
```

Parsing the event can be useful for debugging. We can log the event UUID when we send
a message to the SPM and when we detect an event on the SPM.

#### Host capabilities

Host supporting mailbox events will include

    "mailbox_events": True

in host capabilities response.

#### Upgrade scenarios

When working with both old and new hosts in the same DC, old hosts
will not send events, so the SPM need to continue reading the entire
mailbox every 2 seconds.

Old hosts will read their inbox using larger interval, so extend requests
from old host will take 2-4 seconds.

New hosts will write events and monitor their inbox using lower interval,
so they can have faster extends.

We can change engine to tell the SPM if the DC includes old hosts. The SPM
will read the entire mailbox every 2 seconds only in this case. Otherwise
the SPM can read the entire mailbox only when event was received, or maybe
every 60 seconds as a fallback if something went wrong with events.

### PR DESCRIPTION
**Pull request by nirs**
Minimize message latency by introducing mailbox an events mechanism.

## How this changes system behavior?

The unused host 0 mailbox is used now for sending and receiving mailbox
events.

- Hosts write an event to host 0 mailbox after sending mail to the SPM.

- The SPM monitors host 0 mailbox every eventInterval (0.5 seconds)
  between monitor intervals, so it can handle new messages quickly.

- When hosts wait for reply from the SPM, they monitor their inbox every
  eventInterval (0.5 seconds), so they detect the reply quickly.

- Host reports a new "mailbox_events" capability. This can be used by
  engine to optimize mailbox I/O when all hosts in a data center
  supports this capability.

## Performance improvements

- Extend roundtrip latency reduced from 2.0-4.0 seconds to 0.5-1.0 seconds.
- Total time to extend a volume decreased from 2.3-5.9 seconds to 0.9-3.3
  seconds.
- Maximum write rate before VM starts to paused during extension increased
  from 350 MiB/s to 610 MiB/s. The theoretical improvement is bigger but the
  I could not write faster than 610 MiB/s on the tests system.

## Cost of using events

- The mailbox-spm thread cpu usage increased from 0.3% to 0.6%.
- The additional I/O on the SPM hosts is 4 4k reads every every 2 seconds. (+8k/s).
- On the other hosts, when a host waits for reply to extend message it reads the mailbox (4k) 4 times per second instead of once every 2 seconds. (+14k/s).

## Configuration

To control mailbox events, a new "mailbox:events_enable" was added. This
option allows disabling this optimization in production if needed.

To disable mailbox events, add a drop-in configuration file to all hosts:

    $ cat /etc/vdsm/vdsm.conf.d/99-local.conf
    [mailbox]
    events_enable = false

And restart the vdsmd service.

## Future work

The next improvement is eliminating the wait time. In the worst case, we
waited 1.97 seconds before sending an extend request, which is 68% of the
total extend time (3.31 seconds). This issue is tracked in
https://github.com/oVirt/vdsm/issues/85.

Fixes #102.

### CODE REVIEW

**Review by vjuranek — DISMISSED** [review]

File: lib/vdsm/common/config.py.in [hunk 1]
```diff
@@ -428,6 +428,18 @@ parameters = [
             '(default 20)'),
     ]),
 
+    # Section: [mailbox]
+    ('mailbox', [
+
+        ('events_enable', 'false',
+            'If enabled, hosts use storage events to signal the SPM '
+            'when sending an extend request, and the SPM use storage '
```
Comment (vjuranek) [inline lib/vdsm/common/config.py.in:436]: SPM uses

File: lib/vdsm/common/config.py.in [hunk 2]
```diff
@@ -428,6 +428,18 @@ parameters = [
...
+            'events to detect extend requests quickly. Enabling events '
+            'decreases the time to extend a thin volume and lower the '
```
Comment (vjuranek) [inline lib/vdsm/common/config.py.in:438]: lowers

**Review by vjuranek — APPROVED** [review]

**Comment by nirs** [comment]
My storage cannot write 50g fast enough, but testing 20g disk, I can get even higher
rates without pausing.

guest:

```
# lsblk
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
sda      8:0    0    6G  0 disk 
├─sda1   8:1    0    1M  0 part 
├─sda2   8:2    0    1G  0 part /boot
└─sda3   8:3    0    5G  0 part /
sdb      8:16   0   20G  0 disk 
sr0     11:0    1 1024M  0 rom  
zram0  252:0    0  891M  0 disk [SWAP]

# python3 thinp.py 
20.00 GiB, 33.20 s, 616.88 MiB/s                                               
```

vdsm:

```
2022-03-23 23:24:28,732+0200 INFO  (mailbox-hsm/4) [virt.vm] (vmId='108e7903-08a0-42b7-94cd-2e18272bda06') Extend volume c79cc038-f271-43cf-b6c1-0359306471d1 completed <Clock(total=2.72, wait=1.36, extend-volume=1.14, refresh-volume=0.22)> (thinp:567)
2022-03-23 23:24:32,142+0200 INFO  (mailbox-hsm/0) [virt.vm] (vmId='108e7903-08a0-42b7-94cd-2e18272bda06') Extend volume c79cc038-f271-43cf-b6c1-0359306471d1 completed <Clock(total=2.67, wait=1.89, extend-volume=0.59, refresh-volume=0.18)> (thinp:567)
2022-03-23 23:24:34,641+0200 INFO  (mailbox-hsm/4) [virt.vm] (vmId='108e7903-08a0-42b7-94cd-2e18272bda06') Extend volume c79cc038-f271-43cf-b6c1-0359306471d1 completed <Clock(total=1.56, wait=0.28, extend-volume=1.11, refresh-volume=0.17)> (thinp:567)
2022-03-23 23:24:38,675+0200 INFO  (mailbox-hsm/1) [virt.vm] (vmId='108e7903-08a0-42b7-94cd-2e18272bda06') Extend volume c79cc038-f271-43cf-b6c1-0359306471d1 completed <Clock(total=1.49, wait=0.18, extend-volume=1.11, refresh-volume=0.19)> (thinp:567)
2022-03-23 23:24:42,638+0200 INFO  (mailbox-hsm/2) [virt.vm] (vmId='108e7903-08a0-42b7-94cd-2e18272bda06') Extend volume c79cc038-f271-43cf-b6c1-0359306471d1 completed <Clock(total=1.41, wait=0.13, extend-volume=1.11, refresh-volume=0.17)> (thinp:567)
2022-03-23 23:24:46,239+0200 INFO  (mailbox-hsm/3) [virt.vm] (vmId='108e7903-08a0-42b7-94cd-2e18272bda06') Extend volume c79cc038-f271-43cf-b6c1-0359306471d1 completed <Clock(total=0.92, wait=0.04, extend-volume=0.66, refresh-volume=0.21)> (thinp:567)
2022-03-23 23:24:52,225+0200 INFO  (mailbox-hsm/3) [virt.vm] (vmId='108e7903-08a0-42b7-94cd-2e18272bda06') Extend volume c79cc038-f271-43cf-b6c1-0359306471d1 completed <Clock(total=2.73, wait=1.87, extend-volume=0.60, refresh-volume=0.25)> (thinp:567)
2022-03-23 23:24:56,762+0200 INFO  (mailbox-hsm/1) [virt.vm] (vmId='108e7903-08a0-42b7-94cd-2e18272bda06') Extend volume c79cc038-f271-43cf-b6c1-0359306471d1 completed <Clock(total=2.79, wait=1.40, extend-volume=1.14, refresh-volume=0.25)> (thinp:567)
```

**Comment by michalskrivanek** [comment]
we don't have a bug for this, but please add/link it to https://bugzilla.redhat.com/show_bug.cgi?id=2051997 at least so that we have some "evidence" for improvements

**Comment by nirs** [comment]
> we don't have a bug for this, but please add/link it to https://bugzilla.redhat.com/show_bug.cgi?id=2051997 at least so that we have some "evidence" for improvements

We have upstream issue #102. I can add a comment in the the downstream bug with
some info and link to the upstream bug.

**Comment by nirs** [comment]
Testing cpu usage on the SPM with a real server show that cpu usage of mailbox
thread doing event monitoring is low, around 0.6%.

Tested on:
- model name: Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz
- cores: 80

```
$ top -p 44189 -bHSi -d30 -n10 > with-mailbox-events.top
top - 14:12:15 up  3:50,  2 users,  load average: 0.06, 0.08, 0.08
Threads:  59 total,   0 running,  59 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.1 us,  0.0 sy,  0.0 ni, 99.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 379081.8 free,   3709.5 used,   2518.4 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379274.1 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND

top - 14:12:45 up  3:50,  2 users,  load average: 0.03, 0.07, 0.08
Threads:  59 total,   1 running,  58 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni, 99.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 379081.2 free,   3710.0 used,   2518.4 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379273.5 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  44189 vdsm       0 -20 4841204 135980  30792 S   0.6   0.0   1:10.66 mailbox-spm
  37159 vdsm       0 -20 4841204 135980  30792 S   0.1   0.0   0:53.36 tasks/1
  37160 vdsm       0 -20 4841204 135980  30792 S   0.1   0.0   0:53.47 tasks/2
  37166 vdsm       0 -20 4841204 135980  30792 S   0.1   0.0   0:53.29 tasks/8
  37167 vdsm       0 -20 4841204 135980  30792 S   0.1   0.0   0:57.03 tasks/9
  37172 vdsm       0 -20 4841204 135980  30792 S   0.1   0.0   0:53.31 check/loop
  37175 vdsm       0 -20 4841204 135980  30792 S   0.1   0.0   1:01.16 vdsm.Scheduler
  37209 vdsm       0 -20 4841204 135980  30792 S   0.1   0.0   0:52.26 periodic/2
  37217 vdsm       0 -20 4841204 135980  30792 S   0.1   0.0   0:56.79 health
  43743 vdsm       0 -20 4841204 135980  30792 S   0.1   0.0   0:52.28 monitor/51233f1

top - 14:13:15 up  3:51,  2 users,  load average: 0.02, 0.06, 0.08
Threads:  59 total,   0 running,  59 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni, 99.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 379081.3 free,   3709.9 used,   2518.5 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379273.7 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  44189 vdsm       0 -20 4841204 135984  30792 S   0.5   0.0   1:11.00 mailbox-spm
  37175 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   1:01.38 vdsm.Scheduler
  37164 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.17 tasks/6
  37165 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.31 tasks/7
  37169 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:56.91 mpathhealth
  37172 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.51 check/loop
  37197 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:51.57 jsonrpc/4
  37198 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:51.98 jsonrpc/5
  37208 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:52.36 periodic/1
  43744 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:52.18 monitor/a1b68be

top - 14:13:45 up  3:51,  2 users,  load average: 0.01, 0.05, 0.07
Threads:  59 total,   0 running,  59 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni, 99.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 379081.3 free,   3709.9 used,   2518.5 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379273.7 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  44189 vdsm       0 -20 4841204 135984  30792 S   0.6   0.0   1:11.34 mailbox-spm
  37175 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   1:01.58 vdsm.Scheduler
  37160 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.84 tasks/2
  37166 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.67 tasks/8
  37172 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.70 check/loop
  43743 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:52.66 monitor/51233f1
  43744 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:52.37 monitor/a1b68be
  43766 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:51.87 mailbox-hsm/0
  43769 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:51.72 mailbox-hsm/1
  43772 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:52.06 mailbox-hsm/4

top - 14:14:16 up  3:52,  2 users,  load average: 0.01, 0.05, 0.07
Threads:  59 total,   0 running,  59 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni, 99.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 379081.4 free,   3709.7 used,   2518.5 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379273.8 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  44189 vdsm       0 -20 4841204 135984  30792 S   0.6   0.0   1:11.68 mailbox-spm
  37175 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   1:01.79 vdsm.Scheduler
  43772 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:52.26 mailbox-hsm/4
  37158 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.05 tasks/0
  37159 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.92 tasks/1
  37163 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.45 tasks/5
  37164 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.53 tasks/6
  37165 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.68 tasks/7
  37172 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:53.89 check/loop
  37198 vdsm       0 -20 4841204 135984  30792 S   0.1   0.0   0:52.34 jsonrpc/5

top - 14:14:46 up  3:52,  2 users,  load average: 0.08, 0.06, 0.07
Threads:  60 total,   0 running,  60 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.1 us,  0.1 sy,  0.0 ni, 99.8 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 379057.6 free,   3729.0 used,   2523.0 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379253.4 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  37163 vdsm       0 -20 4841204 136068  30792 S   0.6   0.0   0:54.46 tasks/5
  44189 vdsm       0 -20 4841204 136068  30792 S   0.6   0.0   1:12.69 mailbox-spm
  37198 vdsm       0 -20 4841204 136068  30792 S   0.4   0.0   0:53.30 jsonrpc/5
  37194 vdsm       0 -20 4841204 136068  30792 S   0.1   0.0   0:53.27 jsonrpc/1
  43743 vdsm       0 -20 4841204 136068  30792 S   0.1   0.0   0:53.71 monitor/51233f1
  37160 vdsm       0 -20 4841204 136068  30792 S   0.1   0.0   0:54.87 tasks/2
  37162 vdsm       0 -20 4841204 136068  30792 S   0.1   0.0   0:53.90 tasks/4
  37165 vdsm       0 -20 4841204 136068  30792 S   0.1   0.0   0:54.54 tasks/7
  37166 vdsm       0 -20 4841204 136068  30792 S   0.1   0.0   0:54.70 tasks/8
  37167 vdsm       0 -20 4841204 136068  30792 S   0.1   0.0   0:58.44 tasks/9

top - 14:15:16 up  3:53,  2 users,  load average: 0.11, 0.07, 0.08
Threads:  60 total,   0 running,  60 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.1 us,  0.2 sy,  0.0 ni, 99.5 id,  0.1 wa,  0.0 hi,  0.1 si,  0.0 st
MiB Mem : 385309.7 total, 379075.2 free,   3714.7 used,   2519.7 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379268.6 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  37193 vdsm       0 -20 4841204 136116  30792 S   1.2   0.0   0:54.61 jsonrpc/0
  37198 vdsm       0 -20 4841204 136116  30792 S   0.8   0.0   0:55.09 jsonrpc/5
  44189 vdsm       0 -20 4841204 136116  30792 S   0.6   0.0   1:14.42 mailbox-spm
  37175 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   1:04.24 vdsm.Scheduler
  37196 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:54.91 jsonrpc/3
  37155 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:53.94 logfile
  37163 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.03 tasks/5
  37169 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:59.70 mpathhealth
  37172 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.32 check/loop
  37191 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   1:40.21 JsonRpc (StompR

top - 14:15:46 up  3:53,  2 users,  load average: 0.14, 0.08, 0.08
Threads:  59 total,   0 running,  59 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni, 99.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 379075.8 free,   3714.2 used,   2519.7 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379269.1 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  44189 vdsm       0 -20 4841204 136116  30792 S   0.6   0.0   1:14.82 mailbox-spm
  37175 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   1:04.51 vdsm.Scheduler
  37159 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.58 tasks/1
  37161 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.77 tasks/3
  37163 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.28 tasks/5
  37165 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.35 tasks/7
  37172 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.57 check/loop
  37196 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:55.16 jsonrpc/3
  37200 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:54.23 jsonrpc/7
  37208 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:55.37 periodic/1

top - 14:16:16 up  3:54,  2 users,  load average: 0.09, 0.07, 0.08
Threads:  59 total,   0 running,  59 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni, 99.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 379076.5 free,   3713.4 used,   2519.7 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379269.8 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  44189 vdsm       0 -20 4841204 136116  30792 S   0.5   0.0   1:15.15 mailbox-spm
  37217 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   1:00.21 health
  37175 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   1:04.71 vdsm.Scheduler
  43743 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:55.71 monitor/51233f1
  37159 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.77 tasks/1
  37166 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.69 tasks/8
  37169 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   1:00.13 mpathhealth
  37172 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:56.76 check/loop
  37194 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:55.25 jsonrpc/1
  37207 vdsm       0 -20 4841204 136116  30792 S   0.1   0.0   0:55.77 periodic/0

top - 14:16:46 up  3:54,  2 users,  load average: 0.05, 0.06, 0.07
Threads:  59 total,   0 running,  59 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.1 sy,  0.0 ni, 99.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 379076.8 free,   3713.1 used,   2519.8 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379270.2 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  44189 vdsm       0 -20 4841204 136120  30792 S   0.6   0.0   1:15.68 mailbox-spm
  37165 vdsm       0 -20 4841204 136120  30792 S   0.1   0.0   0:56.92 tasks/7
  43779 vdsm       0 -20 4841204 136120  30792 S   0.1   0.0   0:56.06 mailbox-hsm
  37161 vdsm       0 -20 4841204 136120  30792 S   0.1   0.0   0:57.32 tasks/3
  37163 vdsm       0 -20 4841204 136120  30792 S   0.1   0.0   0:56.84 tasks/5
  37164 vdsm       0 -20 4841204 136120  30792 S   0.1   0.0   0:56.74 tasks/6
  37167 vdsm       0 -20 4841204 136120  30792 S   0.1   0.0   1:00.80 tasks/9
  37172 vdsm       0 -20 4841204 136120  30792 S   0.1   0.0   0:57.14 check/loop
  37175 vdsm       0 -20 4841204 136120  30792 S   0.1   0.0   1:05.09 vdsm.Scheduler
  43744 vdsm       0 -20 4841204 136120  30792 S   0.1   0.0   0:55.77 monitor/a1b68be
```

**Comment by nirs** [comment]
## Testing mailbox events disabled

Withoout mailbox events the mailbox-spm thread consume 0.3% CPU,
but writing more than 350 MiB/s paused once in 20 extends. This
matches what we tested for https://bugzilla.redhat.com/2051997
but it was tested before only on nested environment.

Tested on same server as in https://github.com/oVirt/vdsm/pull/103#issuecomment-1077709954 

```
$ cat /etc/vdsm/vdsm.conf.d/99-local.conf 
[mailbox]
events_enable = false
```

### SPM cpu usage
```
top - 17:52:10 up  7:30,  2 users,  load average: 0.24, 0.19, 0.17
Threads:  59 total,   0 running,  59 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni, 99.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem : 385309.7 total, 376932.2 free,   3748.3 used,   4629.1 buff/cache
MiB Swap:   8192.0 total,   8192.0 free,      0.0 used. 379201.1 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 198711 vdsm       0 -20 4841216 121960  30948 S   0.3   0.0   0:05.87 mailbox-spm
 197984 vdsm       0 -20 4841216 121960  30948 S   0.2   0.0   0:05.58 tasks/0
 198044 vdsm       0 -20 4841216 121960  30948 S   0.2   0.0   0:05.51 health
 198389 vdsm       0 -20 4841216 121960  30948 S   0.2   0.0   0:05.53 monitor/51233f1
 197987 vdsm       0 -20 4841216 121960  30948 S   0.1   0.0   0:05.41 tasks/3
 197989 vdsm       0 -20 4841216 121960  30948 S   0.1   0.0   0:05.94 tasks/5
 197990 vdsm       0 -20 4841216 121960  30948 S   0.1   0.0   0:05.41 tasks/6
 197991 vdsm       0 -20 4841216 121960  30948 S   0.1   0.0   0:05.41 tasks/7
 197993 vdsm       0 -20 4841216 121960  30948 S   0.1   0.0   0:05.54 tasks/9
```

### Extend with 350 MiB/s write

```
# ./extend-stats <without-350.log 

Total time
min=2.510 avg=3.311 max=4.180

Wait time
min=0.230 avg=1.071 max=1.920

Extend time
min=2.030 avg=2.070 max=2.110

Refresh time
min=0.150 avg=0.170 max=0.210
```

Pauses: 0

### Extend with 400 MiB/s writes

```
# ./extend-stats < without-400.log

Total time
min=2.290 avg=3.611 max=5.900

Wait time
min=0.060 avg=0.865 max=1.710

Extend time
min=2.030 avg=2.577 max=4.170

Refresh time
min=0.150 avg=0.170 max=0.190
```

Pauses: 1

```
# grep 'CPU running: onResume' without-400.log 
2022-03-24 18:04:01,520+0000 INFO  (libvirt/events) [virt.vm] (vmId='4b993023-dc51-4a29-8600-1e393dda0f95') CPU running: onResume (vm:6016)
```

**Comment by nirs** [comment]
## Testing mailbox events enabled

With mailbox events the mailbox-spm thread consume 0.6% CPU,
and we cannot paused the VM writing at the maximum rate possible
on this server (610 MiB/s).

Tested on same server as in https://github.com/oVirt/vdsm/pull/103#issuecomment-1077709954

Guest:
```
# python3 thinp.py
50.00 GiB, 83.92 s, 610.07 MiB/s                                              
```

vdsm:
```
# grep 'completed <Clock' vdsm.log | wc -l
100

# ./extend-stats <vdsm.log

Total time
min=0.860 avg=2.001 max=3.270

Wait time
min=0.050 avg=1.005 max=1.990

Extend time
min=0.530 avg=0.819 max=2.080

Refresh time
min=0.150 avg=0.175 max=0.210
```

**Review by bennyz — APPROVED** [review]

**Review by vjuranek — APPROVED** [review]

**Pull request merged**